webserver/webserver/src/test/java/io/helidon/webserver/WaterMarkedBackpressureTest.java

```diff
@@ -0,0 +1,173 @@
+import java.util.logging.Logger;
+
+import io.helidon.common.http.DataChunk;
+import io.helidon.common.reactive.Multi;
+import io.helidon.common.reactive.Single;
+import io.helidon.webclient.WebClient;
+import io.helidon.webclient.WebClientResponse;
+
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import static io.helidon.webserver.BackpressureStrategy.AUTO_FLUSH;
+import static io.helidon.webserver.BackpressureStrategy.LINEAR;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.is;
+
+public class WaterMarkedBackpressureTest {
+
+    private static final Logger LOGGER = Logger.getLogger(WaterMarkedBackpressureTest.class.getName());
+
+    static final Duration TIMEOUT = Duration.ofSeconds(15);
+    private static ExecutorService executorService;
+
+    @BeforeEach
+    void setUp() {
+        executorService = Executors.newSingleThreadExecutor();
+    }
+
+    @AfterEach
+    void tearDown() {
+        executorService.shutdown();
+    }
+
+    @Test
+    void linear() {
+        AtomicLong receivedSize = new AtomicLong(0);
+        AtomicLong lastRequestedNumber = new AtomicLong(0);
+        CompletableFuture<Void> firstBatchSent = new CompletableFuture<>();
+
+        WebServer webServer = null;
+        try {
+            webServer = WebServer.builder()
+                    .host("localhost")
+                    .backpressureBufferSize(500)
+                    .backpressureStrategy(LINEAR)
+                    .routing(r -> r.get("/", (req, res) -> {
+                        res.send(Multi.range(0, 1000)
+                                .observeOn(executorService)
+                                .peek(lastRequestedNumber::set)
+                                // Never flush!
+                                // 5 bytes per chunk
+                                .map(l -> DataChunk.create(false, ByteBuffer.wrap((String.format("%05d", l)).getBytes())))
+                                .onComplete(() -> firstBatchSent.complete(null))
+                                .onCompleteResumeWith(Single.never())
+                        );
+                    }))
+                    .build()
+                    .start()
+                    .await(TIMEOUT);
+
+            WebClient.builder()
+                    .baseUri("http://localhost:" + webServer.port())
+                    .build()
+                    .get()
+                    .path("/")
+                    .request()
+                    .flatMap(WebClientResponse::content)
+                    .forEach(chunk -> {
+                        byte[] bytes = chunk.bytes();
+                        receivedSize.addAndGet(bytes.length);
+                        chunk.release();
+                    })
+                    .onError(Throwable::printStackTrace);
+
+            try {
+                Single.create(firstBatchSent, true).await(Duration.ofSeconds(5));
```

Comment:
If an exception is expected, use `assertThrows` instead.

Reply:
its not expected